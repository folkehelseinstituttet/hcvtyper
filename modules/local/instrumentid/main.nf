process INSTRUMENTID {
    tag "$meta.id"
    label 'process_single'

    // Environment with the seqkit software. Created using seqera containers.
    // URLs:
    // Docker image: https://wave.seqera.io/view/builds/bd-03b4774218b4b7ef_1?_gl=1*1mobsod*_gcl_au*MTkxMjgxNTMwMi4xNzUzNzczOTQz
    // Singularity image: https://wave.seqera.io/view/builds/bd-9a5d37887d7c4e09_1?_gl=1*1mobsod*_gcl_au*MTkxMjgxNTMwMi4xNzUzNzczOTQz
    conda "${moduleDir}/environment.yml"
    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://community-cr-prod.seqera.io/docker/registry/v2/blobs/sha256/90/902e5f138e9145c41c3e7848a9d8b7b7f8a21335ac76ac811b96cabcb3d277ad/data':
        'community.wave.seqera.io/library/seqkit:2.10.0--03b4774218b4b7ef' }"

    input:
    tuple val(meta), path(reads)

    output:
    tuple val(meta), path("*sequencerID.tsv"), emit: id
    path "versions.yml"                      , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"

    """
    # Get the first line of the fastq file
    sqid=\$(seqkit head -n 1 ${reads[0]} | awk 'FNR <= 1')
    echo \${sqid} > ${prefix}.sequencerID.tsv

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        seqkit: \$(seqkit version | cut -d' ' -f2)
    END_VERSIONS
    """

    stub:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"
    """
    # Safe echo (interpolated by Groovy)
    echo "${args}"

    # Create a realistic instrument ID output (mimics seqkit head output)
    echo "@SRR24174266.1 1/1" > ${prefix}.sequencerID.tsv

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        seqkit: \$(seqkit version | cut -d' ' -f2)
    END_VERSIONS
    """
}
