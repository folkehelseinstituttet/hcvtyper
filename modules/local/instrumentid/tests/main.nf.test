nextflow_process {

  name "Test Process INSTRUMENTID"
  script "../main.nf"
  process "INSTRUMENTID"

  tag "modules"
  tag "instrumentid"

  // Real run (uses paired FASTQ files)
  test("instrumentid: paired reads") {
    when {
      process {
        """
        // Input: tuple val(meta), path(reads)
        input[0] = [
          [ id: 'toy' ],
          [
            file("${projectDir}/modules/local/instrumentid/tests/Test_1_R1.fastq.gz", checkIfExists: true),
            file("${projectDir}/modules/local/instrumentid/tests/Test_1_R2.fastq.gz", checkIfExists: true)
          ]
        ]
        """
      }
    }
    then {
      assertAll(
        { assert process.success },  // Check that the process completed successfully (exit code 0)
        // Check that the output files match the expected snapshot
        { assert snapshot([
            // These refer to the process emits
            id      : process.out.id,
            versions: process.out.versions
          ]).match() }
      )
    }
  }

  // Fast stub run
  test("instrumentid: stub") {
    options "-stub-run"
    when {
      process {
        """
        input[0] = [
          [ id: 'toy' ],
          [
            file("${projectDir}/modules/local/instrumentid/tests/Test_1_R1.fastq.gz", checkIfExists: true),
            file("${projectDir}/modules/local/instrumentid/tests/Test_1_R2.fastq.gz", checkIfExists: true)
          ]
        ]
        """
      }
    }
    then {
      assertAll(
        { assert process.success },
        { assert snapshot([
            id      : process.out.id,
            versions: process.out.versions
          ]).match() }
      )
    }
  }
}
