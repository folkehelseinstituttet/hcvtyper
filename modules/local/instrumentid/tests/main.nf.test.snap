{
    "instrumentid: stub": {
        "content": [
            {
                "id": [
                    [
                        {
                            "id": "toy"
                        },
                        "toy.sequencerID.tsv:md5,685fe635d9f55633963e5c04d6f04b91"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,7fd787368f19b6f004808dc74ebeb98c"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.3",
            "nextflow": "25.04.2"
        },
        "timestamp": "2025-10-08T11:30:41.136218685"
    },
    "instrumentid: paired reads": {
        "content": [
            {
                "id": [
                    [
                        {
                            "id": "toy"
                        },
                        "toy.sequencerID.tsv:md5,685fe635d9f55633963e5c04d6f04b91"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,7fd787368f19b6f004808dc74ebeb98c"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.3",
            "nextflow": "25.04.2"
        },
        "timestamp": "2025-10-08T11:29:42.527077154"
    }
}